python/lsst/daf/butler/direct_butler/_direct_butler.py

```diff
@@ -728,16 +747,15 @@ def _rewrite_data_id(
                                 # Look up each visit in the
                                 # visit_system_membership records.
                                 for rec in records:
-                                    membership = list(
-                                        self._registry.queryDimensionRecords(
-                                            # Use bind to allow zero results.
-                                            # This is a fully-specified query.
-                                            "visit_system_membership",
-                                            where="instrument = inst AND visit_system = system AND visit = v",
-                                            bind=dict(
-                                                inst=instrument_records[0].name, system=visit_system, v=rec.id
-                                            ),
-                                        )
+                                    membership = self.query_dimension_records(
+                                        # Use bind to allow zero results.
+                                        # This is a fully-specified query.
+                                        "visit_system_membership",
+                                        where="instrument = inst AND visit_system = system AND visit = v",
```

Comment:
Looks somewhat unfortunate that we have to use string `where` which has to be parsed again slowing things down. I think it would be more efficient to use `Predicate` from new query system instead. But can't we in this particular case use `data_id={"instrument": instrument_records[0].name, "system"=visit_system, "visit"=rec.id}` instead of `where`?

Reply:
Yes. You are right. I will change it over to the data_id since this part is all dimensions (the code above involves dimension metadata). I was also wondering about Predicates but hadn't looked into it (and the SIAv2 code might want to use it as well).

Reply:
Rather than using `Predicate` directly, check out `Query.expression_factory`.

Reply:
Thanks. I'll take a look for dax_obscore, although I might need a bit of help because at the moment I create the WHERE strings and bind parameters in one place when iterating over all the SIAv2 parameters (and have schemes for combining things with OR and AND), and then pass those to the code that is actually inside the Query context. If the expression factory doesn't have to be inside a query context and just needs a universe that will probably be okay.